src/ui.js

```diff
@@ -125,21 +112,70 @@ function handleDragOver(evt) {
 
 
 function buildUploadModal(numFiles) {
-    let getModalContent = numLoaded => `<h1>Reading GPX files...</h1>
-<span id="">${numLoaded} loaded of <b>${numFiles}</b>`;
+    let numLoaded = 0;
+    let failures = [];
+    let getModalContent = () => `
+        <h1>Reading GPX files...</h1>
+        <p>${numLoaded} loaded${
+            failures.length ? `, <span class='failures'>${failures.length} failed</span>` : ``
+        } of <b>${numFiles}</b></p>`;
 
     let modal = picoModal({
-        content: getModalContent(0),
-        closeButton: false,
+        content: getModalContent(),
         escCloses: false,
         overlayClose: false,
         overlayStyles: styles => {
             styles.opacity = 0.1;
         },
-    })
+    });
+
+    modal.afterCreate(() => {
+        // Do not allow the modal to be closed before loading is complete.
+        // PicoModal does not allow for native toggling
+        modal.closeElem().style.display = 'none';
+    });
+
+    modal.afterClose(() => modal.destroy());
 
-    modal.progress = loaded => {
-        modal.modalElem().innerHTML = getModalContent(loaded);
+    // Override the content of the modal, without removing the close button.
+    // PicoModal does not allow for native content overwriting.
+    modal.setContent = body => {
```

Comment:
Possibly a cleaner approach would be to just replace the `innerHTML` of `class="pico-content"`?

Haven't tested this.

Reply:
This is what was happening before, but the close button gets overwritten, as it is a child of `.pico-content`